File: 백준/Silver/15663. N과 M （9）/N과 M （9）.cc
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int N, M;
//vector<int> arr; // 값이 얿을때 비교가 안되고 out of range나와서 일단 배열로 구현
int arr[9] = { 0, };
vector<int> input;
bool isused[9] = { 0, };

void dfs(int k) {
	if (k == M) {
		for (auto i = 0; i < M; ++i) {
			cout << arr[i] << " ";
		}
		cout << "\n";
		arr[k] = 0;
		return;
	}

	for (auto i = 1; i <= N; ++i) {
		if (!isused[i]) {
			if (arr[k] != input[i])arr[k] = input[i];
			else continue;
			isused[i] = 1;
			dfs(k + 1);
			isused[i] = 0;
		}
	}
	arr[k] = 0;
}

int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

	cin >> N >> M;
	input.push_back(0);

	for (auto i = 0; i < N; ++i) {
		int a;
		cin >> a;
		input.push_back(a);
	}

	sort(input.begin(), input.end());
	dfs(0);

	return 0;
}

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[] arr, input;
    static boolean[] vis;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        input = new int[N];
        vis = new boolean[N];
        arr = new int[M];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            input[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(input);

        DFS(0);
        System.out.println(sb);
    }

    static void DFS(int depth) {
        if (depth == M) {
            for (int n : arr) {
                sb.append(n).append(" ");
            }
            sb.append("\n");
            return;
        }

        int temp = -1;
        for (int i = 0; i < N; i++) {
            int now = input[i];
            if (vis[i] || temp == now) {
                continue;
            } else {
                temp = now;
                vis[i] = true;
                arr[depth] = now;
                DFS(depth + 1);
                vis[i] = false;
            }
        }
    }
}
